GraphBase/src/graph/Graph.java

```diff
@@ -0,0 +1,116 @@
+package graph;
+
+import java.util.Map;
+import java.util.HashMap;
+import java.util.NoSuchElementException;
+
+public abstract class Graph {
+
+  protected Map<Integer, Vertex> vertices;
+
+  public Graph() {
+    this.vertices = new HashMap<Integer, Vertex>();
+  }
+
+  public void addVertex(int newVertexId) {
+    Vertex newVertex = new Vertex(newVertexId);
+    vertices.put(newVertex.getId(), newVertex);
+  }
+
+  public void addVertex(Vertex newVertex) {
+    if (vertices.keySet().contains(newVertex.getId())) {
+      System.out.println("The vertex has already been exsited");
+    }
+  }
+
+  public void deleteVertex(int vertexId) {
+    if (!vertices.containsKey(vertexId)) {
+      throw new NoSuchElementException("The vertex id is not existed");
+    }
+    vertices.remove(vertexId);
+    for (Integer key : vertices.keySet()) {
+      if (vertices.get(key).getAdjEdges().keySet().contains(vertexId)) {
+        vertices.get(key).getAdjEdges().remove(vertexId);
+      }
+    }
+  }
+
+  public void deleteVertex(Vertex vertex) {
+    deleteVertex(vertex.getId());
+  }
+
+  public void addEdge(int fromId, int toId) {
+    if (fromId == toId) {
+      System.out.println("Loop is not allowed");
+      return;
+    }
+    if (!vertices.containsKey(fromId)) {
```

Comment:
In addVertex, you will check whether vertices contain the key (as I commented before). So you do not need this code. Simply 
addVertex(fromId);
addVertex(toId);

Reply:
I think this will  way cause to two nesting Method invoking(addVertex(int) addVertex(Vertex) ) . May this reduce the efficiency?

Reply:
No warry about efficiency. We focus more on programming itself. This kind of efficiency loss is nothing.